sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobServiceAsyncClient.java

```diff
@@ -717,4 +740,123 @@ private void throwOnAnonymousAccess() {
                 + "credentials"));
         }
     }
+
+    /**
+     * Restores a previously deleted container.
+     * If the container associated with provided <code>deletedContainerName</code>
+     * already exists, this call will result in a 409 (conflict).
+     * This API is only functional if Container Soft Delete is enabled
+     * for the storage account associated with the container.
+     * For more information, see the
+     * <a href="TBD">Azure Docs</a>. TODO (kasobol-msft) add link to REST API docs
+     *
+     * <p><strong>Code Samples</strong></p>
+     *
+     * {@codesnippet com.azure.storage.blob.BlobServiceAsyncClient.undeleteBlobContainer#String-String}
+     *
+     * @param deletedContainerName The name of the previously deleted container.
+     * @param deletedContainerVersion The version of the previously deleted container.
+     * @return A {@link Mono} containing a {@link BlobContainerAsyncClient} used
+     * to interact with the restored container.
+     */
+    public Mono<BlobContainerAsyncClient> undeleteBlobContainer(
+        String deletedContainerName, String deletedContainerVersion) {
+        return this.undeleteBlobContainer(deletedContainerName,
+            deletedContainerVersion, deletedContainerName);
+    }
+
+    /**
+     * Restores a previously deleted container.  The restored container
+     * will be renamed to the <code>destinationContainerName</code>.
+     * If the container associated with provided <code>destinationContainerName</code>
+     * already exists, this call will result in a 409 (conflict).
+     * This API is only functional if Container Soft Delete is enabled
+     * for the storage account associated with the container.
+     * For more information, see the
+     * <a href="TBD">Azure Docs</a>. TODO (kasobol-msft) add link to REST API docs
+     *
+     * <p><strong>Code Samples</strong></p>
+     *
+     * {@codesnippet com.azure.storage.blob.BlobServiceAsyncClient.undeleteBlobContainer#String-String-String}
+     *
+     * @param deletedContainerName The name of the previously deleted container.
+     * @param deletedContainerVersion The version of the previously deleted container.
+     * @param destinationContainerName The name of the destination container.
+     * @return A {@link Mono} containing a {@link BlobContainerAsyncClient} used
+     * to interact with the restored container.
+     */
+    public Mono<BlobContainerAsyncClient> undeleteBlobContainer(
```

Comment:
I think we can get rid of this overload and the next one. We can have just the min and then the max is WithResponse. 

Also, should we move the destinationContainerName into UndeleteContainerOptions so we don't have to add overloads later. Continuing the conversation we had around tags.

Reply:
Done